Use closefrom if available

nicolas-duteil-nova opened the pull request:
Replaces https://github.com/bellard/quickjs/pull/209

- use `closefrom` if supported
- fallback to the closing loop otherwise
- compute `fd_max` using `/proc/{pid}/fd` on _linux_, if `closefrom` is not supported (ex: musl)

bellard commented:
Since closefrom() is supported in all recent systems, I don't see the advantage in having a complicated fallback using /proc. So I suggest to remove the complicated fallback and just use the loop as fallback.

nicolas-duteil-nova commented:
> Since closefrom() is supported in all recent systems, I don't see the advantage in having a complicated fallback using /proc. So I suggest to remove the complicated fallback and just use the loop as fallback.

> Since closefrom() is supported in all recent systems, I don't see the advantage in having a complicated fallback using /proc. So I suggest to remove the complicated fallback and just use the loop as fallback.

`closefrom()` is not supported in Musl libc for example. The fallback is mostly to deal with cases where `closefrom` is not supported and where relying on `sysconf(_SC_OPEN_MAX);` would impact the performances

nicolas-duteil-nova commented:
What about the last version which uses a separate function to compute the max number of file descriptors and close any descriptor >3 under linux @bellard ?

chqrlie commented:
> What about the last version which uses a separate function to compute the max number of file descriptors and close any descriptor >3 under linux @bellard ?

We are still considering this patch. 
- closefrom should be used when available, including on linux with the musl C library, but we don't have a reliable way to test this special case with adding an extra configuration flag, which we like to avoid.
- the loop could be used as a fallback if `sysconf(_SC_OPEN_MAX)` is large, but not to compute the largest handle, which might still be large in pathological cases, but to actually implement the `closefrom` emulation as done in the other project, while preserving the directory handle. `dirfd` has been available for this purpose on linux for a long time and is POSIX-1.2008 compliant.

Do you have a simple reliable way to test for `closefrom` availability ?

nicolas-duteil-nova commented:
What if we conditionally set `HAVE_CLOSEFROM` in the `Makefile`, after checking that `closefrom` is supported ?

nicolas-duteil-nova commented:
> > What about the last version which uses a separate function to compute the max number of file descriptors and close any descriptor >3 under linux @bellard ?
> 
> We are still considering this patch.
> 
> * closefrom should be used when available, including on linux with the musl C library, but we don't have a reliable way to test this special case with adding an extra configuration flag, which we like to avoid.
> * the loop could be used as a fallback if `sysconf(_SC_OPEN_MAX)` is large, but not to compute the largest handle, which might still be large in pathological cases, but to actually implement the `closefrom` emulation as done in the other project, while preserving the directory handle. `dirfd` has been available for this purpose on linux for a long time and is POSIX-1.2008 compliant.
> 
> Do you have a simple reliable way to test for `closefrom` availability ?

What if we conditionally set HAVE_CLOSEFROM in the Makefile, after checking that closefrom is supported ?

chqrlie commented:
Interesting approach with a Makefile test run on the fly.

I would probably create a subdirectory `compat` for such source files to avoid clutter in the main directory. 
Recompiling this on every invocation including `make clean` is also a bit sloppy, but not necessarily a no-go.

nicolas-duteil-nova commented:
> Interesting approach with a Makefile test run on the fly.
> 
> I would probably create a subdirectory `compat` for such source files to avoid clutter in the main directory. Recompiling this on every invocation including `make clean` is also a bit sloppy, but not necessarily a no-go.

I first try to use a `Makefile` target to do the compilation but I don't think it's possible to define/update a global variable from within a target. Anyway, compiling `test-closefrom.c` is fast enough to no be a pb imho. 

I moved `test-closefrom.c` to `compat` directory and replaced the OS-based support detection with `HAVE_CLOSEFROM`

nickva commented on quickjs-libc.c:
```diff
@@ -2890,6 +2890,57 @@ static int my_execvpe(const char *filename, char **argv, char **envp)
+    and returns the maximum number of open file descriptors on
+    the system.
+ */
+int get_fd_max() {
```
Tiny nit: since it doesn't just get max fds but also closes them sometimes, would it make sense to make the function do all the closing (put `closefrom(4)` here too) and rename it to something like `closefds()`?

nicolas-duteil-nova replied:
> Tiny nit: since it doesn't just get max fds but also closes them sometimes, would it make sense to make the function do all the closing (put `closefrom(4)` here too) and rename it to something like `closefds()`?

Like this ? (https://github.com/bellard/quickjs/pull/239/commits/697650f1d9dffe0a4755c6a7a2ce33dc2976901c)

chqrlie commented:
There is still a problem: you should avoid closing the handle for the directory structure: 

```
                        if (fd > 2 && fd != dirfd(dir)) {
                            close(fd);
                        }
```

nicolas-duteil-nova commented:
@chqrlie : can you approve the workflow ?

nicolas-duteil-nova commented:
@chqrlie : can we merge this one ?

chqrlie commented on quickjs-libc.c:
```diff
@@ -2890,6 +2890,65 @@ static int my_execvpe(const char *filename, char **argv, char **envp)
+    /*
+        Under linux
+
+        - check /proc to find all open file descriptors > 2 and close them
+        - set fd_max = 3 if at least one fd was found in /proc
+
+        This will improve performances on systems where the maximum number
+        of open file descriptor is high (such as in a Docker container).
+    */
+    int pid = getpid();
+    char path[32];
+    struct stat statbuf;
+    snprintf(path, sizeof(path), "/proc/%d/fd", pid);
+    if (stat(path, &statbuf) == 0) {
+        if (S_ISDIR(statbuf.st_mode)) {
+            DIR *dir = opendir(path);
+            if (dir) {
+                struct dirent *subdir;
+                int fd;
+                /*
+                    Close all open file descriptors > 2 and set fd_max to 3
+                */
+                for(;;) {
+                    subdir = readdir(dir);
+                    if (!subdir) {
+                        break;
+                    }
+                    fd = atoi(subdir->d_name);
+                    if (fd > 0) {
+                        fd_max = 3;
+                        if (fd > 2 && fd != dirfd(dir)) {
+                            close(fd);
+                        }
+                    }
+                }
+                closedir(dir);
+            }
+        }
+    }
+#endif
```
Since linux has had `closefrom` for a while now, this approach does not seem useful.

nicolas-duteil-nova replied:
We still need this approach because closefrom is not available on alpine

chqrlie replied:
This configuration setting is indeed system dependant. I have observed these values on some of my systems:
- my laptop running a recent macOS: 256
- Debian linux boxes: 1024 and 65535
- FreeBSD: 28404
- OpenBSD: 512

Modern systems have `closefrom` so we should use that if available and the problem is solved correctly.

On obsolete systems, a best effort approach seems reasonable:
- parsing `/proc/` may actually not be possible as this file system may be disabled.  Furthermore if would only work in linux.
- using `sysconf(_SC_OPEN_MAX)` is reliable but we should limit the loop to a reasonable maximum such as 1024.
- file handles should be open with `O_CLOEXEC` as much as possible.
- it is unlikely more than a few handles will have been opened before `os.exec` is called.
- not closing the handles beyond 1024 on such systems does not have dire consequences.

chqrlie commented on Makefile:
```diff
@@ -119,6 +119,12 @@ ifdef CONFIG_WIN32
 DEFINES+=-D__USE_MINGW_ANSI_STDIO # for standard snprintf behavior
 endif
 
+ifndef CONFIG_WIN32
+ifeq ($(shell $(CC) -o compat/test-closefrom compat/test-closefrom.c 2>/dev/null && echo 1 || echo 0),1)
+DEFINES+=-DHAVE_CLOSEFROM
+endif
+endif
+
```
I am more or less OK with this approach but I have some remarks:
- the executable must not be committed to the repository
- it would be better to invoke the compiler without generating an executable: eg: `clang -fsyntax-only` or to generate the executable to `/dev/null` or at least remove the executable after the check.
- it would be less noisy to add a negative definition `-DCLOSEFROM_NOT_AVAILABLE` if `closefrom` is **not** available, which is not the most common case.

nicolas-duteil-nova replied:
I agree that generating the binary to /dev/null would be an improvement.
But using a negative definition wouldn't change much since we would still need to handle both usecases

nicolas-duteil-nova replied:
Changed the compiler outfile to `/dev/null`.

chqrlie commented on quickjs-libc.c:
```diff
@@ -3026,8 +3084,9 @@ static JSValue js_os_exec(JSContext *ctx, JSValueConst this_val,
             }
         }
 
-        for(i = 3; i < fd_max; i++)
-            close(i);
+        /* close all file descriptors > 2 */
+        close_fds();
+
```
You might expand `close_fds()` inline here

nicolas-duteil-nova replied:
I completely agree with the redirection to `/dev/null` but not sure about the rest. Overall, this PR is an improvement of current behaviour where we loop through all possible fds (which is inefficient). This PR uses `closefrom` when it is available and mitigate the loop problem when `closefrom` is not supported (which is only on alpine afaik).

chqrlie commented:
> @chqrlie : can we merge this one ?

Sorry for the lack of responsiveness, a simpler approach seems advisable, see my proposed changes.

nicolas-duteil-nova commented:
> > @chqrlie : can we merge this one ?
> 
> Sorry for the lack of responsiveness, a simpler approach seems advisable, see my proposed changes.

I completely agree with the redirection to /dev/null but not sure about the rest. Overall, this PR is an improvement of current behaviour where we loop through all possible fds (which is inefficient). This PR uses closefrom when it is available and mitigate the loop problem when closefrom is not supported (which is only on alpine afaik).

chqrlie commented:
> What would be the adverse side effect of merging this PR @chqrlie ?

Well it adds lots of extra code as a partial solution for very specific case that is not so useful.  Fabrice and I try and avoid such additions that may require maintenance and non trivial proofreading. The missing test on `(fd != dirfd(dir))` is a good example of how subtle such things can become.  Parsing the `/proc/` subsystem is a good idea, but not worth the effort in this particular case IMHO.

nicolas-duteil-nova commented:
> > What would be the adverse side effect of merging this PR @chqrlie ?
> 
> Well it adds lots of extra code as a partial solution for very specific case that is not so useful. Fabrice and I try and avoid such additions that may require maintenance and non trivial proofreading. The missing test on `(fd != dirfd(dir))` is a good example of how subtle such things can become. Parsing the `/proc/` subsystem is a good idea, but not worth the effort in this particular case IMHO.

For the record, initial PR had no such pb :grin:

chqrlie commented:
> For the record, initial PR had no such pb 😁

Indeed it did not, the loop was used to determine the largest handle in use, then the directory handle was closed and the loop run from 3 to that handle.

I'm sorry to see you close this PR, I posted a new one using the simplistic approach I suggested, that should fix the problem with alpine linux.
Thank you for your interest in QuickJS.

ctn-malone commented:
> > For the record, initial PR had no such pb 😁
> 
> Indeed it did not, the loop was used to determine the largest handle in use, then the directory handle was closed and the loop run from 3 to that handle.
> 
> I'm sorry to see you close this PR, I posted a new one using the simplistic approach I suggested, that should fix the problem with alpine linux. Thank you for your interest in QuickJS.

No worries @chqrlie. I was just under the impression that we stalled, arguing over details :sweat_smile:. Glad to see that your MR was merged (although I think that detecting the max `fd` through `/proc` would have also handled the cases where the number of `fd` is `> 1024` without any major drawback). Anyway, we all want to make quickjs a better tool which is what matters in the end. Thanks again
